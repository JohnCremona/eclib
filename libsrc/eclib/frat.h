// FRAT.h: interface to FLINT fmpq rational numbers

#if     !defined(_FRAT_H)
#define _FRAT_H      1       //flags that this file has been included

#include "int.h"
#include <flint/fmpq.h>

class RAT {

private:
  fmpq_t q;

public:
  // constructors
  RAT() {
    fmpq_init(q); // sets to 0
  }
  explicit RAT(const INT& n) {
    fmpq_init(q); // sets to 0
    if (n.sign())
      {
        fmpz_t d;
        fmpz_init_set_si(d, 1);
        fmpq_set_fmpz_frac(q, n.z, d);
        fmpz_clear(d);
      }
  }
  RAT(const INT& n, const INT& d) {
    fmpq_init(q); // sets to 0
    fmpq_set_fmpz_frac(q, n.z, d.z);
  }
  explicit RAT(long nu, long de=1) {
    fmpq_init(q); // sets to 0
    fmpq_set_si(q, nu, de);
  }
  RAT(const RAT& x) {
    fmpq_init(q); // sets to 0
    fmpq_set(q, x.q);
  }
  ~RAT() {
    fmpq_clear(q);
  }

  RAT& operator=(const RAT& x) {
    fmpq_init(q); // sets to 0
    fmpq_set(q, x.q);
    return *this;
  }
  RAT operator=(long n) {
    RAT x(n);
    return x;
  }

  // RAT manipulations
  void cancel() {
    fmpq_canonicalise(q); // cancel *this in situ
  }
  INT num() const       // the numerator
  {
    INT n;
    fmpz_init_set(n.z, fmpq_numref(q));
    return n;
  }
  INT den() const        // the denominator
  {
    INT n;
    fmpz_init_set(n.z, fmpq_denref(q));
    return n;
  }
  RAT recip() const  // the reciprocal
  {
    RAT x;
    fmpq_inv(x.q, q);
    return x;
  }

  int sign() const
  {
    return fmpq_sgn(q);
  }

  INT round() const;      // nearest integer

  // Binary Operator Functions
  RAT operator+(const RAT& b) const {
    RAT c;
    fmpq_add(c.q, q, b.q);
    return c;
  }
  RAT operator+(const INT& b) const {
    RAT c;
    fmpq_add_fmpz(c.q, q, b.z);
    return c;
  }
  RAT operator+(long b) const {
    RAT c;
    fmpq_add_si(c.q, q, b);
    return c;
  }
  RAT operator+(int b) const {
    RAT c;
    fmpq_add_si(c.q, q, b);
    return c;
  }
  friend inline RAT operator+(const INT& a, const RAT& b) {return b+a;}
  friend inline RAT operator+(long a, const RAT& b) {return b+a;}
  friend inline RAT operator+( int a, const RAT& b) {return b+a;}
  RAT operator-(const RAT& b) const {
    RAT c;
    fmpq_sub(c.q, q, b.q);
    return c;
  }
  RAT operator-(INT b) const {
    RAT c;
    fmpq_sub_fmpz(c.q, q, b.z);
    return c;
  }
  RAT operator-(long b) const {
    RAT c;
    fmpq_sub_si(c.q, q, b);
    return c;
  }
  RAT operator-(int b) const {
    RAT c;
    fmpq_sub_si(c.q, q, b);
    return c;
  }
  friend inline RAT operator-(const INT& a, const RAT& b) {return -(b-a);}
  friend inline RAT operator-(long a, const RAT& b) {return -(b-a);}
  friend inline RAT operator-(int a, const RAT& b) {return -(b-a);}

  RAT operator*(const RAT& b) const {
    RAT c;
    fmpq_mul(c.q, q, b.q);
    return c;
  }
  RAT operator*(const INT& b) const {
    RAT c;
    fmpq_mul_fmpz(c.q, q, b.z);
    return c;
  }
  RAT operator*(long b) const {
    RAT c;
    fmpq_mul_si(c.q, q, b);
    return c;
  }
  RAT operator*(int b) const {
    RAT c;
    fmpq_mul_si(c.q, q, b);
    return c;
  }
  friend inline RAT operator*(const INT& a, const RAT& b) {return b*a;}
  friend inline RAT operator*(const long a, const RAT& b) {return b*a;}
  friend inline RAT operator*(const int a, const RAT& b) {return b*a;}

  RAT operator/(const RAT& b) const {
    RAT c;
    fmpq_div(c.q, q, b.q);
    return c;
  }
  RAT operator/(const INT& b) const {
    RAT c;
    fmpq_div_fmpz(c.q, q, b.z);
    return c;
  }
  RAT operator/(long b) const {
    RAT c;
    fmpz_t z;
    fmpz_init_set_si(z, b);
    fmpq_div_fmpz(c.q, q, z);
    fmpz_clear(z);
    return c;
  }
  RAT operator/(int b) const {
    RAT c;
    fmpz_t z;
    fmpz_init_set_si(z, b);
    fmpq_div_fmpz(c.q, q, z);
    fmpz_clear(z);
    return c;
  }
  friend inline RAT operator/(const INT& a, const RAT& b) {
    RAT c = b/a;
    return c.recip();
  }
  int operator==(const RAT& b) const {
    return fmpq_equal(q, b.q);
  }
  int operator==(INT b) {
    return fmpq_equal_fmpz(q, b.z);
  }
  int operator==(long b) {
    return fmpq_equal_si(q, b);
  }
  friend inline int operator==(INT a, RAT b) {return b==a;}
  friend inline int operator==(long a, RAT b) {return b==a;}
  friend inline int operator!=(RAT a, RAT b) {return !(a==b);}
  friend inline int operator!=(RAT a, INT b) {return !(a==b);}
  friend inline int operator!=(INT a, RAT b) {return b!=a;}
  friend inline int operator!=(long a, const RAT& b) {return b!=INT(a);}

  int operator<(const RAT& b) const {return fmpq_cmp(q, b.q)<0;}
  int operator<=(const RAT& b) const {return fmpq_cmp(q, b.q)<=0;}
  int operator<(const INT& b) const {return fmpq_cmp_fmpz(q, b.z)<0;}
  int operator<=(const INT& b) const {return fmpq_cmp_fmpz(q, b.z)<=0;}
  int operator<(long b) const {return fmpq_cmp_si(q, b)<0;}
  int operator<=(long b) const {return fmpq_cmp_si(q, b)<=0;}
  int operator>(const RAT& b) const {return fmpq_cmp(q, b.q)>0;}
  int operator>=(const RAT& b) const {return fmpq_cmp(q, b.q)>=0;}
  int operator>(const INT& b) const {return fmpq_cmp_fmpz(q, b.z)>0;}
  int operator>=(const INT& b) const {return fmpq_cmp_fmpz(q, b.z)>=0;}
  int operator>(long b) const {return fmpq_cmp_si(q, b)>0;}
  int operator>=(long b) const {return fmpq_cmp_si(q, b)>=0;}
  friend inline int operator<(const INT& a, const RAT& b) {return b>a;}
  friend inline int operator<=(const INT& a, const RAT& b) {return b>=a;}
  friend inline int operator<(long a, const RAT& b) {return b>a;}
  friend inline int operator<=(long a, const RAT& b) {return b>=a;}
  friend inline int operator>(const INT& a, const RAT& b) {return b<a;}
  friend inline int operator>=(const INT& a, const RAT& b) {return b<=a;}
  friend inline int operator>(long a, const RAT& b) {return b<a;}
  friend inline int operator>=(long a, const RAT& b) {return b<=a;}

  friend inline RAT max(const RAT& a, const RAT& b);
  friend inline RAT min(const RAT& a, const RAT& b);

  friend ostream& operator<< (ostream&, const RAT&);
  friend istream& operator>> (istream&, RAT&);

  void operator+=(const RAT& b) {fmpq_add(q, q, b.q);}
  void operator+=(const INT& b) {fmpq_add_fmpz(q, q, b.z);}
  void operator+=(int b) {fmpq_add_si(q, q, b);}
  void operator-=(const RAT& b) {fmpq_sub(q, q, b.q);}
  void operator-=(const INT& b) {fmpq_sub_fmpz(q, q, b.z);}
  void operator-=(int b) {fmpq_sub_si(q, q, b);}
  void operator*=(const RAT& b) {fmpq_mul(q, q, b.q);}
  void operator*=(const INT& b) {fmpq_mul_fmpz(q, q, b.z);}
  void operator*=(int b) {fmpq_mul_si(q, q, b);}
  void operator/=(const RAT& b) {fmpq_div(q, q, b.q);}
  void operator/=(const INT& b) {fmpq_div_fmpz(q, q, b.z);}
  void operator/=(int b) {
    fmpz_t bb;
    fmpz_init_set_si(bb,b);
    fmpq_div_fmpz(q, q, bb);
    fmpz_clear(bb);
  }

  RAT operator+() const {
    RAT x(*this);
    return x;
  }
  RAT operator-() const {
    RAT x(*this);
    fmpq_neg(x.q, x.q);
    return x;
  }
  RAT abs() const {
    RAT x(*this);
    fmpq_abs(x.q, x.q);
    return x;
  }
  INT floor() const;
  INT ceil() const;
  int is_square() const {
    return (num()*den()).is_square();
  }
  int is_square(RAT& a) const {
    INT nd = num()*den(), rnd;
    int b=nd.is_square(rnd);
    if (b) a = RAT(rnd, den());
    return b;
  }
};


// Definitions of non-member RAT functions


inline INT RAT::round() const {
  return rounded_division(num(), den(), 0); // 0 means halves go up
}

inline ostream& operator<<(ostream& s, const RAT& q)
{
  INT n(q.num()), d(q.den());
  if (is_zero(d))
    s<<"oo";
  else
    {
      s << n;
      if (d!=1)
        s << "/" << d;
    }
  return s;
}

inline istream& operator>> (istream& is, RAT& r)
{
  std::string n;
  is>>n;
  fmpq_set_str(r.q, n.c_str(), 10);
  fmpq_canonicalise(r.q);
  return is;
}

inline INT RAT::floor() const
{
  INT q, r;
  divrem(num(), den(), q, r);
  return q;
}

inline INT RAT::ceil() const
{
  INT q, r;
  divrem(num(), den(), q, r);
  return (is_zero(r)? q : q+1);
}

inline int sign(const RAT& r)
{
  return r.sign();
}

inline RAT max(const RAT& a, const RAT& b) {return (a>=b? a : b);}
inline RAT min(const RAT& a, const RAT& b) {return (a<=b? a : b);}


#endif
