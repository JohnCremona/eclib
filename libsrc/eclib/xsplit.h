// xsplit.h: Declaration of class form_finder
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
//                     Marcus Mo     (parallel code)
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#if     !defined(_ECLIB_XSPLIT_H)
#define _ECLIB_XSPLIT_H      1       //flags that this file has been included

// Disable multithreading
// #undef ECLIB_MULTITHREAD

#ifdef ECLIB_MULTITHREAD 
#include <boost/thread/mutex.hpp>
#endif

#include <eclib/xsplit_data.h>
#include <eclib/splitbase.h>
#ifdef ECLIB_MULTITHREAD
#include <eclib/threadpool.h>
#endif

// flags set on construction:

// dual:  use dual bases throughout (the usual case)
// plusflag: if 1 then targetdim=1 and conjmat not used
//           if 0 then targetdim=2 and conjmat is used

// bigmats: if 1 then uses opmat() to get operators on ambient space,
// does its own restriction to the current subspace; if 0 then uses
// opmat_restricted() to get operators restricted to current space
// directly -- ONLY possible when dual=1

class form_finder {
  public:
    form_finder(splitter_base* hh, int plus, int maxd, int mind=0, 
                int dualflag=1, int bigmatsflag=0, int v=0);
    ~form_finder(void); 
    
    void find();
    void find(ff_data &data);
    void recover(vector< vector<long> > eigs);
    void splitoff(const vector<long>& eigs);
    void store(vec bp, vec bm, vector<long> eigs);
    
    vec  getbasis( const ff_data &data ) const {return data.bplus_;}
    vec  getbasisplus( const ff_data &data ) const {return data.bplus_;}
    vec  getbasisminus( const ff_data &data ) const {return data.bminus_;}

    friend class ff_data; 
  
  protected:
    splitter_base* h;
    
    int            plusflag, dual, bigmats, verbose, targetdim;
    int            gnfcount;                  // Global newform counter
    long           maxdepth, mindepth, dimen;
    SCALAR         denom1;
    vector< vector<long> > gaplist;           // Vector to hold all (sub)eiglists
    vector<vec>    gbplus, gbminus;           // Vector to hold all bplus/bminus

    ff_data* root;                            // Always points to root data node

    void make_opmat(long i, ff_data &data);   // Puts it in the_opmat
    void make_submat(ff_data &data);
    smat make_nested_submat(long ip, ff_data &data);
    void go_down(ff_data &data, long eig, int last=0);
    void go_up( ff_data &data );
    void make_basis(ff_data &data);
    vec make_basis1(ff_data &data);  // current space has dim. 1
    vec make_basis2(ff_data &data, const vec& v);  // current space has dim. >1 and
                                                   // relative basis vector v

#ifdef ECLIB_MULTITHREAD
    threadpool   pool;                        // Job queue
    boost::mutex store_lock;                  // Lock for store() function
#endif
};

vec  getbasis1(const ssubspace* s);       // Assuming dim(s)=1, get basis vector
vec  lift(const vec& v);                  // Lift basis vector

#endif
