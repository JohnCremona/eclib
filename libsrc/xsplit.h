// xsplit.h: Declaration of class form_finder
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#if     !defined(_XSPLIT_H)
#define _XSPLIT_H      1       //flags that this file has been included

#include "method.h"  // #defines form_finder=form_finder0/1/2/3/4
#include "splitbase.h"

// flags set on construction:

// dual:  use dual bases throughout (the usual case)
// plusflag: if 1 then targetdim=1 and conjmat not used
//           if 0 then targetdim=2 and conjmat is used

// bigmats: if 1 then uses opmat() to get operators on ambient space,
// does its own restriction to the current subspace; if 0 then uses
// opmat_restricted() to get operators restricted to current space
// directly -- ONLY possible when dual=1

class form_finder {
protected:
  splitter_base* h;
  int plusflag, dual, bigmats, verbose, targetdim;
  long maxdepth, mindepth, depth, subdim, dimen;
  SCALAR denom1;
  ssubspace** nest;     // array of pointers to subspaces
// "Current" subspace is *nest[depth] of dimension subdim
  vector<long> eiglist;
  vec bplus, bminus;

  int *havemat;
  char** opfilenames;  // temp filenames
  smat conjmat;  // only used if plus==0 and bigmats==1
  smat the_opmat;
  smat *submats;  // holds current restriction for i>0
  void make_opmat(long i);  // puts it in the_opmat
  void make_submat();
  void go_down(long eig, int last=0);
  void go_up();
  void make_basis();
  vec getbasis1(const ssubspace* s); //assuming dim(s)=1, get basis vector
public:
  form_finder(splitter_base* hh, int plus, int maxd, int mind=0, 
              int dualflag=1, int bigmatsflag=0, int v=0);
  ~form_finder(void); 
  void find();
  void recover(vector< vector<long> > eigs);
  void splitoff(const vector<long>& eigs);
  vec getbasis() const {return bplus;}
  vec getbasisplus() const {return bplus;}
  vec getbasisminus() const {return bminus;}
};

#endif
