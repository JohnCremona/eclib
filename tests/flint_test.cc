#include "eclib/xsplit.h"   // which includes linalg.h
#include "eclib/frat.h"

void testINT(INT a)
{
  cout<<"a = "<<a;
  cout<<", -a = "<<-a;
  cout<<", a+20 = "<<a+20;
  cout<<", 20+a = "<<20+a;
  cout<<", a-2 = "<<a-2;
  cout<<", 2-a = "<<2-a;
  cout<<", a.abs() = "<<a.abs();
  cout<<", sign(a) = "<<sign(a);
  cout<<endl;
  INT b = (a^8) - 1;
  cout<<"b = a^4-1 = "<<b;
  cout<<", prime factors: "<<pdivs(b);
  cout<<endl;
}

int main ()
{
  INT a(-13), b(3), c;
  cout << "a = "<<a<<" with sign "<<sign(a)<<endl;
  cout << "b = "<<b<<" with sign "<<b.sign()<<endl;
  cout << "c = "<<c<<" with sign "<<sign(c)<<endl;
  cout << "-a = "<<-a<<" with sign "<<sign(-a)<<endl;
  cout << "a.abs() = "<<a.abs()<<endl;
  cout << "abs(b) = "<<abs(b)<<endl;
  c = a+b;
  cout << "a+b = "<<c<<endl;
  cout << "a-b = "<<a-b<<endl;
  cout << "a*b = "<<a*b<<endl;
  cout << "2+a = "<<2+a<<endl;
  cout << "a+2 = "<<a+2<<endl;
  cout << "2-a = "<<2-a<<endl;
  cout << "a-2 = "<<a-2<<endl;
  cout << "2*a = "<<2*a<<endl;
  cout << "a*2 = "<<a*2<<endl;
  cout << "a%b = "<<a%b<<endl;
  cout << "mod(a,b) = "<<mod(a,b)<<endl;
  cout << "posmod(a,b) = "<<posmod(a,b)<<endl;
  cout << "gcd(a,b) = "<<gcd(a,b)<<endl;
  testINT(a);
  testINT(b);
  INT x, y;
  INT g = bezout(a,b,x,y);
  cout << "bezout(a,b,x,y) = "<<g<<" with x="<<x<<", y="<<y<<endl;
  int e=9;
  INT p = a^e;
  cout << "a^"<<e<<" = "<<p<<endl;
  long f = 3;
  cout << "b^"<<f<<" = "<<(b^f)<<endl;
  cout << "a==b? " << (a==b) << endl;
  cout << "a!=b? " << (a!=b) << endl;
  cout << "a==-13? " << (a==-13) << endl;
  cout << "a!=-13? " << (a!=-13) << endl;
  cout << "a==0? " << (a==0) << endl;
  cout << "a!=0? " << (a!=0) << endl;
  INT n = a*(a+1)*((a^2)+1);
  cout<<"Prime factors of "<<n<<" : "<<pdivs(n) << endl;
  cout<<"Divisors whose square divides "<<n<<" : "<<sqdivs(n) << endl;
  INT pr(7);
  for (long i=0; i<pr; i++)
    cout<<"("<<i<<"|"<<pr<<") = "<<legendre(INT(i),pr)<<endl;
  cout<<"Is "<<a<<" square? "<<a.is_square()<<endl;
  INT a2 = a*a;
  cout<<"Is "<<a2<<" square? "<<a2.is_square()<<endl;
  INT s = a2.isqrt();
  cout<<"(a square root is "<<s<<")"<<endl;
  cout<<"Enter an integer: "<<flush;
  cin >> a;
  cout<<"   value entered: "<<a<<endl;
  if (a.is_long())
    {
      cout<<" as a long int: "<<I2long(a)<<endl;
    }
  else
    {
      cout<<" does not fit in a long int!"<<endl;
    }

  RAT half(1,2);
  cout << "one half = " << half <<", rounds to "<<half.round()<<", floor="<<half.floor()<<", ceil="<<half.ceil()<<", recip="<<half.recip()<<endl;
  RAT third(1,3);
  cout << "one third = " << third <<", rounds to "<<third.round()<<", floor="<<third.floor()<<", ceil="<<third.ceil()<<", recip="<<third.recip()<<endl;

  cout<<"Enter a rational: "<<flush;
  RAT q;
  cin >> q;
  cout<<"   value entered: "<<q<<endl;
}
