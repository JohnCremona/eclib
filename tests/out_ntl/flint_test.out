a = -13 with sign -1
b = 3 with sign 1
c = 0 with sign 0
-a = 13 with sign 1
a.abs() = 13
abs(b) = 3
a+b = -10
a-b = -16
a*b = -39
2+a = -11
a+2 = -11
2-a = 15
a-2 = -15
2*a = -26
a*2 = -26
a%b = 2
mod(a,b) = -1
posmod(a,b) = 2
gcd(a,b) = 1
a = -13, -a = 13, a+20 = 7, 20+a = 7, a-2 = -15, 2-a = 15, a.abs() = 13, sign(a) = -1
b = a^4-1 = 815730720, prime factors: [ 2 3 5 7 17 14281 ]
a = 3, -a = -3, a+20 = 23, 20+a = 23, a-2 = 1, 2-a = -1, a.abs() = 3, sign(a) = 1
b = a^4-1 = 6560, prime factors: [ 2 5 41 ]
bezout(a,b,x,y) = 1 with x=-1, y=-4
a^9 = -10604499373
b^3 = 27
a==b? 0
a!=b? 1
a==-13? 1
a!=-13? 0
a==0? 0
a!=0? 1
Prime factors of 26520 : [ 2 3 5 13 17 ]
Divisors whose square divides 26520 : [ 1 2 ]
(0|7) = 0
(1|7) = 1
(2|7) = 1
(3|7) = -1
(4|7) = 1
(5|7) = -1
(6|7) = -1
Is -13 square? 0
Is 169 square? 1
(a square root is 13)
Enter an integer:    value entered: 1234567890
 as a long int: 1234567890
one half = 1/2, rounds to 0, floor=0, ceil=1, recip=2
one third = 1/3, rounds to 0, floor=0, ceil=1, recip=3
Enter a rational:    value entered: 1000000000000000000000000/300000000000001
